crates/scheduler/src/scheduling/batch_scheduler.rs

```diff
@@ -35,153 +88,372 @@ pub enum BatchSchedulerError {
-/// flowchart TD
-/// A[Training] --> |Worker: Status| B{Project}
-/// B --> |Not done| A
-/// B -->|Done| C[UpdateScheduled]
-/// C -->|Worker: Update| E[Updating]
-/// E -->|PS: Updated| F[Update Received]
-/// F --> A
-/// ```
+/// Handle action protocol requests and respond with next steps.
 async fn schedule<T, S>(
     tx: Sender<(PeerId, Metrics)>,
-    tracker: Arc<Mutex<ProgressTracker<T>>>,
-    request: (PeerId, progress::Request),
-    notifier: Arc<Notify>,
-) -> Result<progress::Response, BatchSchedulerError>
+    worker_pool: PoolStatisticsHandle<T>,
+    parameter_pool: PoolHandle,
+    round_state: Arc<Mutex<RoundState>>,
+    training_state: Arc<Mutex<TrainingState>>,
+    batch_sizer: BatchSizer,
+    start: std::time::Instant,
+    request: (PeerId, action::ActionRequest),
+) -> Result<action::ActionResponse, BatchSchedulerError>
 where
     T: RuntimeStatistic + 'static,
     S: Simulation + Send + Sync + 'static,
 {
-    let (
-        peer_id,
-        progress::Request {
-            progress: update, ..
-        },
-    ) = request;
-    tracing::debug!(
-        %peer_id,
-        ?update,
-        "Received update",
-    );
-    match update {
-        progress::Progress::Metrics(metrics) => {
-            tx.send((peer_id, metrics))
-                .await
-                .map_err(BatchSchedulerError::from)?;
-            Ok(progress::Response::Ok {})
-        }
-        progress::Progress::Status(progress::Status { batch_size }) => {
-            let mut tracker = tracker.lock().await;
-            tracker.update(&peer_id, batch_size)?;
-            match tracker.worker_tracker.worker_state(&peer_id) {
-                State::Training => {
-                    // NOTE: time_cap is u64 (time), update_cap is u32 (count)
-                    let time_cap: u64 = 10000;
-                    let update_cap: u32 = 3;
+    let _ = std::marker::PhantomData::<S>;
+    let (peer_id, action::ActionRequest { job_id, status }) = request;
+    tracing::debug!(%peer_id, ?status, %job_id, "Received action request");
+
+    let now = SystemTime::now();
+
+    let since_start = start.elapsed().as_millis() as u64;
+    // NOTE: We rely on Pool::members() being oldest-first ordered by join time.
+    let parameter_servers: Vec<PeerId> =
+        parameter_pool.members().iter().map(|w| w.peer_id).collect();
+    let primary_ps = parameter_servers.first().copied();
+
+    let response = match status {
+        ExecutorStatus::Train(train) => match train {
+            TrainStatus::Idle => ExecutorAction::Train(TrainAction::ExecuteBatch),
+            TrainStatus::BatchCompleted { batch_size } => {
+                worker_pool.update(&peer_id, since_start);
+
+                let snapshot = worker_pool.statistics();
+                let mut training = training_state.lock().await;
+                training.record_batch(batch_size);
+                let samples_remaining = training.samples_remaining();
+                drop(training);
```

Comment:
Nit: At least for me it's more readable to use a separate block instead of using `drop`.

```suggestion
                let samples_remaining = {
                  let mut training = training_state.lock().await;
                  training.record_batch(batch_size);
                  training.samples_remaining()
                };
```

Reply:
Yep, 100%!